cli/src/main/scala/com/lightbend/rp/reactivecli/docker/DockerRegistry.scala

```diff
@@ -0,0 +1,142 @@
+          image = image,
+          tag = providedTag.getOrElse(DockerDefaultTag),
+          providedUrl = providedUrl,
+          providedNamespace = providedNs,
+          providedImage = image,
+          providedTag = providedTag))
+  }
+
+  def getBlob(uri: String, digest: String, token: Option[String]): Try[(HttpResponse, Option[String])] =
+    for {
+      i <- parseImageUri(uri)
+      r <- getWithToken(blobUrl(i, digest), HttpHeaders(Map.empty), token = token)
+    } yield r
+
+  def getConfig(uri: String, token: Option[String]): Try[(Config, Option[String])] =
+    for {
+      manifest <- getManifest(uri, token)
+      blob <- getBlob(uri, manifest._1.config.digest, token = manifest._2)
+      config <- getDecoded[Config](blob._1)
+    } yield config -> blob._2
+
+  def getDecoded[T](response: HttpResponse)(implicit decode: DecodeJson[T]): Try[T] =
+    if (response.statusCode == 200)
+      response.body.getOrElse("").decodeEither[T].fold(
+        err => Failure(new IllegalArgumentException(s"Decode Failure: $err")),
+        Success.apply)
+    else
+      Failure(new IllegalArgumentException(s"Expected code 200, received ${response.statusCode}"))
+
+  def getManifest(uri: String, token: Option[String]): Try[(Manifest, Option[String])] =
+    for {
+      i <- parseImageUri(uri)
+      r <- getWithToken(manifestUrl(i), HttpHeaders(Map("Accept" -> DockerAcceptManifestHeader)), token = token)
+      v <- getDecoded[Manifest](r._1)
+    } yield v -> r._2
+
+  private def getWithToken(url: String, headers: HttpHeaders, tryNewToken: Boolean = true, token: Option[String] = None): Try[(HttpResponse, Option[String])] = {
+    val request =
+      HttpRequest(url)
+        .headers(token.fold(headers)(t => headers.updated("Authorization", s"Bearer $t")))
+        .enableFollowRedirects
+
+    LibHttpSimple(request).flatMap {
+      case response if response.statusCode == 401 && response.headers.contains("Www-Authenticate") && tryNewToken =>
+        logger.debug("Received 401, attempting to get a token and try again")
+
+        val authenticateHeader = response.headers("Www-Authenticate")
+
+        val maybeResponse =
+          for {
+            realm <- parseWwwAuthenticate(authenticateHeader, "realm")
+            service <- parseWwwAuthenticate(authenticateHeader, "service")
+            scope <- parseWwwAuthenticate(authenticateHeader, "scope")
+            tokenResponse <- LibHttpSimple(HttpRequest(tokenUrl(realm, service, scope, "LightbendReactiveCLI"))).toOption
+
+            if tokenResponse.statusCode == 200
+
+            body <- tokenResponse.body
+            json <- JsonParser.parse(body).right.toOption
+            token <- json.field("token")
+            tokenStr <- token.string
+          } yield getWithToken(url, headers, tryNewToken = false, token = Some(tokenStr))
+
+        maybeResponse.getOrElse {
+          logger.error("Unable to obtain an OAuth token")
```

Comment:
Perhaps also log the HTTP response code and response text as part of the error message?